### PR TITLE
[UI Apps] Auth State

### PR DESCRIPTION
**Pull request by enbashi**
Follow the read me here: https://github.com/DataDog/apps-auth-sample/blob/main/README.md

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -195,3 +196,22 @@ export type GetDashboardCogMenuItemsRequest = RequireKeys<
...
+export interface AuthStateOptions {
+    url: string;
+    closePopupAfterAuth?: boolean;
+    authStateCallback: () =>
+        | Promise<CustomAuthState | boolean>
+        | CustomAuthState
+        | boolean;
+    retryInterval?: number;
+    totalTimeout?: number;
+}
```
Comment (nmuldavin) [inline src/types.ts:217]: I think we should obscure some of these options from the user unless users explicitly ask for control over these things. Doing so will give us more flexibility over the implementation details while we figure things out. I'd say we can remove `retryInterval`, `totalTimeout`, `closePopupAfterAuth` and keep the other two

**Review by nmuldavin — COMMENTED** [review]
I'm leaving my main feedback for the work across the repos in this PR so that it's centralized.

This works well, thanks in particular for the very thorough demo app!

The polling of `authStateCallback` while the authorization popup is open is really unfortunate. I do understand why it's implemented that way, it is the easiest way to maintain the constraint that we can use existing application authentication mechanisms with zero effort required from developers.

It would be better all around if there were a way for developers to 'tell us' when the auth state is resolved. One idea i had is to make a generic redirect page that developers could redirect to after the user authenticates, like`/apps/resolve-auth-state?app_id=id&isAuthenticated=true&...`. It would be a blank page with a spinner that would maybe be running framepost to send a message to the parent saying the auth state is resolved before closing itself. 

That would mean that the developers would need some of their auth logic to be specific to the datadog app, at least the handler that says to redirect to that url on successful auth. I wonder if the UX + performance improvements would be worth that minimal effort?

Potentially it could be an option we give developers called like `resolutionType= "poll" | "redirect"` They could do the easiest thing with polling to get off the ground, but we could recommend a redirect resolution as the better way to do it if they want to put the minimal effort in.

I left some other comments about naming and where we put certain logic inline below, happy to discuss!

File: src/types.ts (hunk 1 above)
Comment (enbashi) [reply src/types.ts:217]: They are optional with defaults so they are there if needed. I think some power users would want to fine tune the experience depending on the how expensive/performance the underlying logic in their `authStateCallback`. At least one customer already requested to close the popup themselves so I think providing the flexibly without forcing it is good.

File: src/auth/auth.ts [hunk 2]
```diff
@@ -1,33 +1,139 @@
...
+        this.updateAuthState({ status: AuthStateStatus.INITIATED });
+
+        await this.client.framePostClient.request(
+            UiAppRequestType.AUTH_WITH_POPUP_INIT,
+            {
+                authUrl: this.authStateProvider!.options.url
+            }
         );
+
+        return new Promise(resolve => {
```
[the remark this answers is not in the document]
Comment (enbashi) [reply src/auth/auth.ts:71]: We can look into that once we add session persistence to share state across frames. For now, I think having one the SDK instantiated per app per frame gives us scoping for free without having to map auth state to app is in web-ui. I'd probably require your other app state PR to be merged first. We can reiterate then.

File: src/auth/auth.ts [hunk 3]
```diff
@@ -1,33 +1,139 @@
...
+        return this.authState;
     }
 
-    // Pass the URL query params. Must be called in a popup window opened by requestAuthTokens()
-    // The counterpart to requestAuthTokens()
-    resolveAuthTokens(paramsString: string) {
-        this.client.framePostClient.onRequest(
-            UiAppRequestType.REQUEST_AUTH_TOKENS,
-            () => paramsString
+    async authenticateWithPopup(): Promise<AuthState> {
```
[the remark this answers is not in the document]
Comment (enbashi) [reply src/auth/auth.ts:50]: ok for `setAuthStateProvider` -> `setProvider`. I see us adding other ways in the future to initiate the flow like modals, side panels or even inline for cases where iframes work fine so `authenticateWithPopup` is better be explicit.

**Comment by enbashi** [comment]
> I'm leaving my main feedback for the work across the repos in this PR so that it's centralized.
> 
> This works well, thanks in particular for the very thorough demo app!
> 
> The polling of `authStateCallback` while the authorization popup is open is really unfortunate. I do understand why it's implemented that way, it is the easiest way to maintain the constraint that we can use existing application authentication mechanisms with zero effort required from developers.
> 
> It would be better all around if there were a way for developers to 'tell us' when the auth state is resolved. One idea i had is to make a generic redirect page that developers could redirect to after the user authenticates, like`/apps/resolve-auth-state?app_id=id&isAuthenticated=true&...`. It would be a blank page with a spinner that would maybe be running framepost to send a message to the parent saying the auth state is resolved before closing itself.
> 
> That would mean that the developers would need some of their auth logic to be specific to the datadog app, at least the handler that says to redirect to that url on successful auth. I wonder if the UX + performance improvements would be worth that minimal effort?
> 
> Potentially it could be an option we give developers called like `resolutionType= "poll" | "redirect"` They could do the easiest thing with polling to get off the ground, but we could recommend a redirect resolution as the better way to do it if they want to put the minimal effort in.
> 
> I left some other comments about naming and where we put certain logic inline below, happy to discuss!

I think a generic redirect page might work, but it will require a some tedious configuration from the developer's side because the full url will change based on the env (us, eu, etc) and most auth providers require the redirect url to be predefined. Until we support cross-region availability, I don't think it's worth the effort. 

I think the redirect use case can be simpler to handle with the second auth provider that I demoed earlier. Basically the redirect page will be part of this app itself and we leverage the [framepost's requestChannelWithPopup](https://github.com/DataDog/framepost/blob/a93f58f4677ef400a66d69879e91d9526bd8c870/src/parent.ts#L30) that I added it just for this purpose so we can run the sdk on the popup. Developers can then just call `client.auth.resolveAuthState()` on demand without polling.  (a working POC is still available [here](https://github.com/DataDog/apps-auth-sample/commit/e817c87699cc9be78d8bd07e6a1aabb01cdfb5fe#diff-fb4c0964cee9fff342f2a735cd559ee2e53f096495ce2ea2b71fe923315d7c55R15-R22)).

That said, I believe the simplicity of polling approach might be appealing to most developers specially with the ability to customize the logic in `authStateCallback` to short circuit it based on the existence of a localStorage token or cookie and/or fine tune timeout/interval periods for expensive API calls. I'd rather stick to the simplicity of this approach and get some feedback from users before considering a more complex solution.

File: src/auth/auth.ts [hunk 4]
```diff
@@ -1,33 +1,139 @@
...
+                const newAuthState = {
+                    status: AuthStateStatus.FAILED,
+                    isAuthenticated: false
+                };
+
+                this.updateAuthState(newAuthState);
+                resolve(newAuthState);
+            }, this.authStateProvider!.options.totalTimeout);
+
+            const interval = setInterval(async () => {
```
[the remark this answers is not in the document]
Comment (enbashi) [reply src/auth/auth.ts:84]: I can see if I can detect if window is closed. We can also lower the default timeout period.

File: src/types.ts (hunk 1 above)
Comment (nmuldavin) [reply src/types.ts:217]: I'm explicitly thinking about what will happen if we do implement a way for the developer to tell us that the auth state is resolved without polling, like my redirect idea or a different method. In that case `retryInterval` and `totalTimeout` might not apply anymore. `closePopupAfterAuth` won't apply if we choose a different auth method besides a popup. Keep in mind we can always give them control over those settings if they ask for it ... going the opposite way is harder

**Comment by nmuldavin** [comment]
yeah your original POC using framepost and `client.auth.resolveAuthState()` would work really well actually, that is a much better idea than mine :) I agree polling is the simplest approach for now. Let's roll out with it, and investigate other methods as a followup

File: src/auth/auth.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmuldavin) [reply src/auth/auth.ts:50]: If we end up supporting different authentication methods like modals + sidepanels I think that should be a field on the provider options like:
```
setProvider({
  url: '',
  method: 'modal', // defaults to 'popup',
  ...

})
```

The advantage is that once defined, we can call a single `authenticate` handler from anywhere, including in web-ui, based on the provider definition. For example, we should probably make authentication state a first class thing and add an empty state to custom widgets like this (generated by editing the dom): 
Screen Shot 2021-04-07 at 8 21 40 AM

When the user clicks that button, we know how the app wants to authenticate based on the provider definition, then we can launch the flow on their behalf.

**Review by nmuldavin — APPROVED** [review]

**Pull request merged**